sktime/annotation/base/_base.py

```diff
@@ -314,3 +345,231 @@ def _update(self, X, Y=None):
         self._fit(self._X, self._Y)
 
         return self
+
+    def predict_segments(self, X):
+        """Predict segments on test/deployment data.
+
+        Parameters
+        ----------
+        X : pd.DataFrame
+            Data to annotate, time series.
+
+        Returns
+        -------
+        Y : np.ndarray
+            2D array containing the segments for X. The first column contains the labels
```

Comment:
this should clarify whether you mean `loc` or `iloc`.

Further, I think it needs to be able to represent ends as well.

The format I've used in private code is `pd.DataFrame`, with columns `seg_start` and `seg_end`, and optional `seg_label`.

Noting that, to compress, you could remove:

* `seg_end`, if segments are consecutive and exhaustive
* `seg_label`, if segments are unlabelled

There is no way I see in the `np.ndarray` case to map the four cases you can have (having or not having either of the two) on different shapes (there are only three cases if you have 3 cols max).

Reply:
Are you okay with the following?
* dense segments being a series amd sparse segments being a dataframe.
* sparse and dense change points being a series.
* sparse and dense anomalies being a series.

Using series avoids having to have different column names for segmentation, anomaly detection, and change point detection. I think it makes the code cleaner.

> this should clarify whether you mean loc or iloc

I haven't included this in the docstrings yet but I would prefer `iloc` for the time being. I think unordered indexes could cause issues `loc` and sparse segmentation. I think we could get around this with some proper type checking but in the interest of time, I would rather leave this for the time being.

Reply:
> Using series avoids having to have different column names for segmentation, anomaly detection, and change point detection. I think it makes the code cleaner.

hmm - how do you represent segments that are not "tiled", i.e., the union is not the entire range, or they do overlap?

Example:
* original time series (to segment) has indices 1, 2, 3, 4, 5
* segments go [1, 2] and [4, 5]

Example 2:
* same time series
* segments go [1, 3] and [2, 4]

Reply:
> I think unordered indexes could cause issues `loc` and sparse segmentation.

So, points "between indices" is not allowed?

Why this could make sense: let's say you are using the anomaly detection metric which is the distance to the closest anomaly plus a cost per anomaly that you detect.

This is in general optimized by predicting it to be somewhere in-between the indices, rather than restricting it to observed indices only.

Reply:
(feel free to ignore any of the above, if it leads to "unclear design" issues)

Reply:
Thanks for raising these points.

> the union is not the entire range.

In sparse format I don't think that is too much of problem. Indexes that are not covered by the segments are unclassified.

In dense format, I suggest we reserve an integer to represent unclassified segments. I think `-1` would make the most sense since `0` is reserved is used to represent none changepoints/anomalies.

Here, I check for zeroes. https://github.com/Alex-JG3/sktime/blob/e7a748671488a7d945bc8a2095fb4bf9e6edc125/sktime/annotation/base/_base.py#L538.

> or they do overlap?

Currently, it is assumed the segments do not overlap. However, in sparse format I don't think is too difficult to represent. Your example would be dataframe that looks like (I have taken 1 of the indexes to account for iloc instead of loc):

```python
   seg_label  seg_start  seg_end
0          1          0        1
1          2          2        3
```

Converting this into dense is more difficult. I would suggest we use a dataframe where each column represents a different label. Your example becomes:

```python
   label_1  label_2
0        1        0
1        1        0
2        0        0
3        0        1
4        0        1
```

It would be nice deal with this in a way without so many zeros. But, I don't think lists in cells is well supported in pandas.

> So, points "between indices" is not allowed?

Not currently but I see your point.

This sounds especially relevent if the timeseries is not uniformly sampled. For anomalies in between indexes, I would say we need `loc` instead of `iloc`. I think this particular point adds quite a bit of complexity that would be better handled carefully in its own PR. I think the other points can be dealt with now.

Reply:
> Your example would be dataframe that looks like (I have taken 1 of the indexes to account for iloc instead of loc):

Yes, that makes sense. `seg_end` could be optional and default to the `seg_start` of the next row - though in this case one would have to worry about the end of the last one. So not sure if this is fully thought out.

> Converting this into dense is more difficult. I would suggest we use a dataframe where each column represents a different label. Your example becomes:

Hm, that's one-hot encoded.
As long as they do not overlap, you could still have unique label, and 0 or -1 for no label?

There's pros/cons for adopting a 1D representation as long as it is possible. Con is that you have to switch representations if not; Pro is that 1D is useable more universally (e.g., as a target with an estimator that needs 1D `y`)

> I think this particular point adds quite a bit of complexity that would be better handled carefully in its own PR.

Agreed.

Reply:
PS: have you seen `pandas.IntervalIndex`?

It can be used as a column type:
https://pandas.pydata.org/docs/reference/api/pandas.IntervalIndex.html

Not arguing that we use it, as it may come with its own issues.

There is an interesting parallel, the `from_breaks` method, related to the discussion above.

Reply:
> have you seen `pandas.IntervalIndex`?

I havn't is that `is_overlapping` function looks like it would be useful.